subcommand/acl-controller/command.go

```diff
@@ -185,6 +295,131 @@ var partitionedClientPolicyTpl = `partition "%s" {
   }
 }`
 
+// upsertConsulClientRole creates or updates the Consul ACL role for the client token.
+func (c *Command) upsertConsulClientRole(consulClient *api.Client, roleName, policyName string) error {
+	if err := c.upsertClientPolicy(consulClient, policyName); err != nil {
+		return err
+	}
+	if err := c.upsertClientRole(consulClient, roleName, policyName); err != nil {
+		return err
+	}
+	return nil
+}
+
+// upsertClientPolicy creates the ACL policy for the Consul client, if the policy does not exist.
+func (c *Command) upsertClientPolicy(consulClient *api.Client, policyName string) error {
+	// If the policy already exists, we're done.
+	_, _, err := consulClient.ACL().PolicyReadByName(policyName, c.queryOptions())
+	if err != nil && !controller.IsACLNotFoundError(err) {
+		return fmt.Errorf("reading Consul client ACL policy: %w", err)
+	} else if err == nil {
+		c.log.Info("ACL policy already exists; skipping policy creation", "name", policyName)
+		return nil
+	}
+
+	// Otherwise, the policy is not found, so create it.
+	c.log.Info("creating ACL policy", "name", policyName)
+	rules := ossClientPolicy
+	if c.flagPartitionsEnabled {
+		// If partitions are enabled then create a policy that supports partitions
+		rules = fmt.Sprintf(partitionedClientPolicyTpl, c.flagPartition)
+	}
+	_, _, err = consulClient.ACL().PolicyCreate(&api.ACLPolicy{
+		Name:        policyName,
+		Description: "Consul Client Token Policy for ECS",
+		Rules:       rules,
+	}, c.writeOptions())
+	if err != nil {
+		return fmt.Errorf("creating Consul client ACL policy: %w", err)
+	}
+	c.log.Info("ACL policy created successfully", "name", policyName)
+	return nil
+}
+
+// upsertClientRole creates the ACL role for the Consul client, if the role does not exist.
+func (c *Command) upsertClientRole(consulClient *api.Client, roleName, policyName string) error {
+	// If the role already exists, we're done.
+	role, _, err := consulClient.ACL().RoleReadByName(roleName, c.queryOptions())
+	if err != nil && !controller.IsACLNotFoundError(err) {
+		return fmt.Errorf("reading Consul client ACL role: %w", err)
+	} else if err == nil && role != nil { // returns role=nil and err=nil if not found
+		c.log.Info("ACL role already exists; skipping role creation", "name", roleName)
+
+		if len(role.Policies) == 0 {
+			c.log.Info("updating ACL role with policy", "role", roleName, "policy", policyName)
+			role.Policies = []*api.ACLLink{{Name: policyName}}
+			_, _, err := consulClient.ACL().RoleUpdate(role, c.writeOptions())
+			if err != nil {
+				return fmt.Errorf("updating Consul client ACL role: %s", err)
+			}
+			c.log.Info("update ACL role successfully", "name", roleName)
+		}
```

Comment:
When would this happen? It looks like policies are included on the role [in a single operation](https://github.com/hashicorp/consul-ecs/pull/71/files#diff-57537ae8c46b9830b6b0ab7933cce6705bf85d9f419f741bcb216178c306f702R361) below.

Reply:
Yeah, this a rare what-if scenario. It would happen if the policy is deleted somehow (manually?), then the link to the policy is removed from the role. The way I wrote the unit test made it easy to test for this, so I just handled it.